Progress report - February 2021 (#263)
yuzu-emu/yuzu-emu.github.io: site/content/entry/yuzu-progress-report-feb-2021/index.md

AMD users still require to install the latest `Optional` driver version to get support for `VK_EXT_robustness2`. At the time of writing, the latest version is `21.2.3`, but yuzu will work with drivers as old as `20.12.1`.

Multi-GPU systems must have all their drivers updated, even for integrated graphics that are functional but not in use.

## Technical bugfixes

[Merry](https://github.com/MerryMage) recently caught a bug in the implementation of yuzu's `SPSC` ring buffer and fixed it by [removing the granularity template argument](https://github.com/yuzu-emu/yuzu/pull/5885).
What does this mean exactly? Let's explain!

A buffer is a data structure that reserves space in memory as slots to store information temporarily: for example, an audio buffer.
In particular, a [ring buffer](https://en.wikipedia.org/wiki/Circular_buffer) is a special type of buffer where the slot next to the final one is the first slot in the buffer (so the start and the end are connected).
Once it's full, no new data is added until some information has been extracted from the buffer and processed.

> `SPSC` stands for "Single-Producer-Single-Consumer," and is a model that comes from the [Producer-Consumer problem](https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem) proposed by computer scientists to deal with the problem of proper synchronization when various simultaneous processes write and read from the same buffer.

Reviewer: Agree with the change (can't commit). I will let flam push the change with his review.

One of the many tasks of the kernel is to assign resources to processes whenever they ask for them.
For this reason, [epicboy](https://github.com/ameerj) started the work necessary to [utilize a more accurate resource_limit implementation](https://github.com/yuzu-emu/yuzu/pull/5877), in order to match the hardware behaviour more closely.

Be it memory, threads, or ports, the kernel checks for their availability and keeps track of them through a variable called `resource limit`.
By comparing the current amount of resources being used against the resource limit, the kernel can determine whether to deny a request or not.
This stems from the fact that resources are finite, especially in weaker hardware such as that in the Nintendo Switch.
A PC, on the other hand, isn't as restricted as a Nintendo Switch.
Until now, whenever a process requested resources, yuzu would create its own instance of `resource limit` instead of using a system-wide variable to keep track of it.
This PR is just the initial step in preparation to reverse engineer the correct behaviour and implement it in the emulator.

> ## Paint me like one of your french bits

Reviewer: It's intended to be funny more than accurate. I figured the reference would click on most people once they start reading about how the GPU can basically be used to operate over information (bits) even though it might not necessarily be textures or images. And yeah, I considered pixels, but really it's hard to think of something different from bit that fits as nicely here haha
Will let flam decide over this one.

The reason for this design choice is simply because parallelisation is a very efficient scheme to process computer graphics. Since a single instruction is capable of operating over many components of data at the same time, such as the vertices and textures of a 3D scene, this increases the throughput of information, especially when compared against the performance of running the same operations on a CPU.
Their potential isn't limited to just these functions though.
It is possible to write programs that won't necessarily operate over graphics, yet still benefit from these characteristics.
This is known as `GPGPU` - [General-purpose computing on graphics processing units](https://en.wikipedia.org/wiki/General-purpose_computing_on_graphics_processing_units) - and it's intended to be used when there is a problem that can be separated into a number of parallel tasks in order to be processed more efficiently. These problems are commonly called `embarrassingly parallel problems`.

One of these cases was the [use of compute shaders to decode ASTC textures](https://github.com/yuzu-emu/yuzu/pull/5927).

> `ASTC` stands for "Adaptable Scalable Texture Compression," and it's a fairly new image compression format mainly aimed at mobile devices.

Reviewer: Yeah, why not.

The Nintendo Switch is capable of decoding these textures natively in hardware, but it's a feature that most PC GPU vendors lack in their products (with the exception of Intel Graphics, being the only vendor that offers native support).

> The decoding of these textures is therefore a non-trivial task that can have a huge impact on performance, as seen in games such as `Astral Chain` and `Luigi's Mansion 3`.

Reviewer: What about

> The decoding of these textures is therefore a non-trivial task that can have a huge impact on performance. Two notable examples are `Astral Chain` and `Luigi's Mansion 3`, since both games make extensive use of this format, but it can also be observed to various degrees in other titles, where these textures are generally used to a lesser degree in menu icons, minimaps, etc.

> The decoding of these textures is therefore a non-trivial task that can have a huge impact on performance, as seen in games such as `Astral Chain` and `Luigi's Mansion 3`.

Reviewer 2: perfect!

{{< imgs
	"./astral_chain_atsc.mp4| Comparison between the old and the new implementation of the ASTC decoder"
  >}}

> This led to the implementation of an `ASTC` decoder through the CPU, which was faster than what GPUs could do with their lack of support, but was still far from being a satisfactory solution since it consumed CPU resources and consequently slowed down games that made extensive use of this format.

Reviewer: Agree with the change myself, but since it involves rewording I will leave it up to flam.

This way, the load on the CPU will be shifted to the GPU, allowing emulation to run in parallel with the texture decoding.
As a side benefit, now textures remain in the GPU memory all the time, since they don't need to be transferred between CPU and GPU for decoding.
This means that there won't be time spent downloading the texture to the CPU and then uploading it back to the GPU after the decoding is done, like in the old implementation.

This feature works as intended on all GPU vendors on Windows, although there are a few problems on Linux (more specifically, the `AMDGPU-PRO` driver) that still need to be ironed out.
Our devs are working hard to solve these bugs, so we ask our tux-friends to be patient and stay tuned!

Since compute programs were originally meant to manipulate image data, they also worked out nicely to fix a problem with one of the rendering APIs used in yuzu: by [using compute shaders to swizzle BGR textures on copy](https://github.com/yuzu-emu/yuzu/pull/5891).

{{< single-title-imgs
    "Color-swapped and properly swizzled versions of Octopath Traveler's title screen"
    "./octopath1.jpg"
    "./octopath2.jpg"
  >}}

> In OpenGL, colours are stored in channels, and the way they are laid out varies depending on the format used.

Reviewer 2: ```suggestion
In OpenGL, colors are stored in channels, and the way they are laid out varies depending on the format used.
```
are you a brit?

Reviewer: Will probably be needed to find the other references and change them too.

For example, the `RGB` format stores the color channels in the order "Red, Green and, Blue," while the `BGR` format stores the channels in the order "Blue, Green, and Red".
Unfortunately, this latter format isn't supported internally by OpenGL, which caused problems with a number of games that made use of `BGR` textures: their Red and Blue channels were swapped and the final images looked blue-ish.

{{< single-title-imgs
    "Blue Christina looks nice, but the red hue definitely suits the Nixie Tubes much better in Steins;Gate: My Darling's Embrace"
    "./sg1.jpg"
    "./sg2.jpg"
  >}}

> The solution to this problem then was to reorder the Blue and Red channels in the copy uploaded into the GPU.

Reviewer 2: ```suggestion
The solution to this problem then was to reorder the Blue and Red channels of `BGR` textures in the copy uploaded to the GPU.
```

Reordering the graphical information of an image to process it in the graphics card is called swizzling, so what this PR does is copy the values of the Red channel into the Blue channel and vice-versa, a process that can be exploited through parallel computation.

> This way, the problem with OpenGL is directly bypassed on the GPU, and games can render as they should on the screen.

Reviewer 2: ```suggestion
This way, the limitation with OpenGL is directly bypassed on the GPU, and these textures are rendered as intended on screen.
```

## General bug fixes and improvements

`Pokémon Sword and Shield` players can enjoy one less frequent crash. 
Boss [bunnei](https://github.com/bunnei) [fixed a problem on LDN initialization](https://github.com/yuzu-emu/yuzu/pull/5920), eliminating the crash that occurred if the player pressed `Y` during gameplay (activating online services that yuzu lacks).
An error window will still pop up, but emulation will continue.

> Yet another `Animal Crossing: New Horizons` update, yet another service to stub or implement to regain playability.

Reviewer 2: ```suggestion
Another `Animal Crossing: New Horizons` update, another service to stub or implement to regain playability.
```
"yet" seems off in this sentence imo
